src/whylogs/core/metrics/confusion_matrix.py

```diff
@@ -0,0 +1,148 @@
+from typing import List, Union
+
+import numpy as np
+from sklearn.utils.multiclass import type_of_target
+
+from whylogs.proto import ScoreMatrixMessage
+from whylogs.core.statistics import NumberTracker
+
+
+SUPPORTED_TYPES = ("binary", "multiclass")
+
+
+class ConfusionMatrix:
+
+    """
+    Confusion Matrix Class to hold labels and matrix data
+
+    Attributes:
+        confusion_matrix (nd.array): Description
+        labels (List[str]): Description
+    """
+
+    def __init__(self, labels: List[str] = None,
+                 prediction_field=None,
+                 target_field=None,
+                 score_field=None):
+        self.prediction_field = prediction_field
+        self.target_field = target_field
+        self.score_field = score_field
+        if labels:
+            self.labels = sorted(labels)
+            num_labels = len(self.labels)
+            self.confusion_matrix = np.empty(
+                [num_labels, num_labels], dtype=object)
+            for each_ind_i in range(num_labels):
+                for each_ind_j in range(num_labels):
+                    self.confusion_matrix[each_ind_i,
+                                          each_ind_j] = NumberTracker()
+        else:
+            self.labels = None
+            self.confusion_matrix = None
+
+    def add(self, predictions: List[Union[str, int, bool]],
+            targets: List[Union[str, int, bool]],
+            scores: List[float]):
+        """
+        add. predictions and targets to confusion matrix with scores
+
+        Args:
+            predictions (TYPE): Description
+            targets (TYPE): Description
+            scores (TYPE): Description
+
+        Raises:
+            ValueError: Description
+        """
+
+        tgt_type = type_of_target(targets)
+        if tgt_type not in ("binary", "multiclass"):
+            raise NotImplementedError("target type not supported yet")
+
+        if not isinstance(targets, list):
+            targets = [targets]
+        if not isinstance(predictions, list):
+            predictions = [predictions]
+
+        if scores is None:
+            scores = [1.0 for _ in range(len(targets))]
+
+        if len(targets) != len(predictions):
+            raise ValueError(
+                "both targets and predictions need to have the same length")
+
+        targets_indx = enconde_to_integers(targets, self.labels)
+        prediction_indx = enconde_to_integers(predictions, self.labels)
+
+        for ind in range(len(predictions)):
+            self.confusion_matrix[prediction_indx[ind],
```

Comment:
FYI Java implementation will rebuild the matrix if it detects a new label.

Reply:
Also you probably want to handle the case when `prediction_indx[ind] == -1` here

Reply:
cant happen :)  since the `enconde_to_integers()` is strictly 0->N

Reply:
Okay apparently you get exception in this:

```
enconde_to_integers(["1"], ["2", "3"])
```

Result:

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-8-30f02dc386b1> in <module>
----> 1 enconde_to_integers(["1"], ["2", "3"])

<ipython-input-5-e9e1d6398927> in enconde_to_integers(values, uniques)
      1 def enconde_to_integers(values, uniques):
      2     table = {val: i for i, val in enumerate(uniques)}
----> 3     return np.array([table[v] for v in values])

<ipython-input-5-e9e1d6398927> in <listcomp>(.0)
      1 def enconde_to_integers(values, uniques):
      2     table = {val: i for i, val in enumerate(uniques)}
----> 3     return np.array([table[v] for v in values])

KeyError: '1'
```

Would be good to have more meaningful error message (`values not specified in labels` for example)

Reply:
yes but `enconde_to_integers()` is only used with the joined labels of prediction and target, or with combined labels of two confusions matrices. so that branch doesnt exist

Reply:
Gotcha. It's worth to note that that this is internal class and is not meant to used outside of ModelProfile